### ISSUE TITLE
readDocument Google Cloud Vision in Cloud Functions does not work

### ISSUE THREAD

**Issue by rafaelvizcaino** [GoogleCloudPlatform/cloud-functions-emulator#102]
I need to write a Cloud Function, and the sole permitted programming language is Node. Honestly I just have a slight idea about Node, but I am daring to write code based on some examples given in the GCP documentation. Here it is:

```js
'use strict';

// [START functions_ocr_setup]
const config = require('./config.json');

// Get a reference to the Cloud Storage component
const storage = require('@google-cloud/storage')({
  keyFilename: './pathtomyserviceaccount.json',
  projectId: 'projectId'
});

// Get a reference to the Cloud Vision API component
const vision = require('@google-cloud/vision')({
  keyFilename: './pathtomyserviceaccount.json',
  projectId: 'projectId'
});

// [END functions_ocr_setup]

// [START functions_ocr_detect]
function detectText (file) {
  console.log('Haya vamos!');

  return vision.readDocument(file);
    .then((results) => {
      const fullTextAnnotation = results[1].responses[0].fullTextAnnotation;
      console.log(fullTextAnnotation.text);
    })
    .catch((err) => {
      console.error('ERRORes:', err);
    });
}
// [END functions_ocr_detect]

// [START functions_ocr_process]
/**
 * Cloud Function triggered by Cloud Storage when a file is uploaded.
 *
 * @param {object} event The Cloud Functions event.
 * @param {object} event.data A Google Cloud Storage File object.
*/
exports.processImage = function processImage (event) {
  let file = event.data;

  return Promise.resolve()
    .then(() => {
      if (file.resourceState === 'not_exists') {
        // This was a deletion event, we don't want to process this
        return;
      }

      if (!file.bucket) {
        throw new Error('Bucket not provided. Make sure you have a "bucket" property in your request');
      }
      if (!file.name) {
        throw new Error('Filename not provided. Make sure you have a "name" property in your request');
      }

      return detectText(storage.bucket(file.bucket).file(file.name));
    })
    .then(() => {
      console.log(File ${file.name} processed.);
    });
};
// [END functions_ocr_process]
```

The file 'serviceaccount.json' is located in the same folder as the index.js (yes, it contains the function 'processImage' that I want GC to execute).

Looking in the log of the corresponding GC Funtion, I get the following error:

> PartialFailureError: A failure occurred during this request

I've been doing some testing, and I intuit that the error is related to the 'readDocument' call... Once again, I recognize I'm a newcomer in Node, so I would really appreciate any hint or suggestion!!! Many thanks in advance!!!

**Comment by jmdobry** [comment]
Instead of `vision.readDocument(file)`, try this:

```js
vision.readDocument(storage.bucket(file.bucket).file(file.name))
```

Also, instead of passing `keyFilename` and `projectId` to the client library instantiation, it's better to set the `GOOGLE_APPLICATION_CREDENTIALS` and `GCLOUD_PROJECT` environment variables, e.g.

```
export GOOGLE_APPLICATION_CREDENTIALS=/path/to/my/serviceaccount.json
export GCLOUD_PROJECT=my-project-id
```

then in your code just do:

```js
const storage = require('@google-cloud/storage')();
const vision = require('@google-cloud/vision')();
```

**Comment by rafaelvizcaino** [comment]
Many thanks for your fast reply! I followed the next steps:

First of all, I tested the correction `vision.readDocument(storage.bucket(file.bucket).file(file.name))` without changing anything else in the code... and it is giving me the same error. Therefore, since the readDocument parameter is now well introduced (thanks to your advice) I suppose that the problem is to point correctly at the project and to give correctly the service account.

Secondly, I tried to set the environment variables. I have been surfing for a while (in fact, I found this: https://github.com/GoogleCloudPlatform/nodejs-docs-samples/issues/117) in order to get an idea of what to do before asking, but I am in a loss. Could you please give me any documentation to set environment variables for a GC Function?

Many thanks for your kind help!

**Comment by jmdobry** [comment]
1. Set the `GOOGLE_APPLICATION_CREDENTIALS` environment variable.

        export GOOGLE_APPLICATION_CREDENTIALS=/path/to/my/serviceaccount.json

1. Set the `GCLOUD_PROJECT` environment variable.

        export GCLOUD_PROJECT=my-project-id

1. Start the Emulator:

        functions start

1. Move into your function directory:

        cd /path/to/function/dir

1. Deploy your function:

        functions deploy processImage --trigger-bucket your-image-bucket

1. Call your function

        functions call processImage --data '{"bucket":"your-image-bucket","name":"your-file.txt"}'

**Comment by rafaelvizcaino** [comment]
First of all, thanks again for your kind help! Now I understand you are trying to explain me what to do in the case I were testing my code in my local machine.

What I'm doing is deploying directly my app into GC and testing the Cloud Function on it. Afterwards, I get into the Stackdriver log with the aim of seeing what has happened.

So let me rephrase the question:

I'm currently testing the following code:

```
'use strict';

// [START functions_ocr_setup]
const config = require('./config.json');

const opt={
  keyFilename: './myserviceaccountkey.json',
  projectId: 'my-project-id'
};

// Get a reference to the Cloud Storage component
const storage = require('@google-cloud/storage')(opt);

// Get a reference to the Cloud Vision API component
const vision = require('@google-cloud/vision')(opt);

// [END functions_ocr_setup]

// [START functions_ocr_detect]

function detectText (file) {
	
  console.log('Haya vamos!');
  
  return vision.readDocument(storage.bucket(file.bucket).file(file.name))
    
	.then((results) => {
		const fullTextAnnotation = results[1].responses[0].fullTextAnnotation;
		console.log(fullTextAnnotation.text);
	})
	.catch((err) => {
		console.error('ERRORes:', err);
	});
	
}

// [END functions_ocr_detect]

// [START functions_ocr_process]
/**
 * Cloud Function triggered by Cloud Storage when a file is uploaded.
 *
 * @param {object} event The Cloud Functions event.
 * @param {object} event.data A Google Cloud Storage File object.
 */
exports.processImage = function processImage (event) {
  
	let file = event.data;
	
	return Promise.resolve()
	
	.then(() => {
      if (file.resourceState === 'not_exists') {
        // This was a deletion event, we don't want to process this
        return;
      }

      if (!file.bucket) {
        throw new Error('Bucket not provided. Make sure you have a "bucket" property in your request');
      }
      if (!file.name) {
        throw new Error('Filename not provided. Make sure you have a "name" property in your request');
      }
	  
      return detectText(storage.bucket(file.bucket).file(file.name));
    })
    .then(() => {
      console.log(`File ${file.name} processed.`);
    });
	
};
// [END functions_ocr_process]
```

...and the log is:

![log](https://cloud.githubusercontent.com/assets/28631206/26262690/ea63d7fc-3cd5-11e7-9aec-37725c2739da.png)

Taking all these circumstances into consideration... what is your opinion? Could it be possible that I have some problem with the service account key for Vision? I am at complete loss...

Many thanks in advance!

**Comment by jasonpolites** [comment]
Not sure if it's related, but the path you use for loading files from the local file system should use the `__dirname` global.  There's no guarantee that the working directory is the same as the deployed directory.  `require` might be ok, but I suspect the `keyFilename` may not be.

```javascript
const path = require('path');
let keyFilename = path.resolve(__dirname, 'myserviceaccountkey.json');
```
Secondly, if you're accessing the API enabled in the *same project* as the Cloud Function, you shouldn't need to bother with the keyfile at all.  As @jmdobry has suggested:

```javascript
const storage = require('@google-cloud/storage')();
const vision = require('@google-cloud/vision')();
```

This should pick up the service account that the function itself is *running as*

**Comment by rafaelvizcaino** [comment]
Many thanks for your fast reply @jasonpolites 

Yes, the API is enabled in the same project as the Cloud Function. Everything is in the same project, I just work with one project. Despite this, it does not work. I'm in the same situation I have described one day ago.

So currently I am using the following code:

```
'use strict';

// [START functions_ocr_setup]
const config = require('./config.json');

// Get a reference to the Cloud Storage component
const storage = require('@google-cloud/storage')();

// Get a reference to the Cloud Vision API component
const vision = require('@google-cloud/vision')();

// [END functions_ocr_setup]

// [START functions_ocr_detect]

function detectText (file) {
	
  console.log('Haya vamos!');
  
  return vision.readDocument(file)
    
	.then((results) => {
		const fullTextAnnotation = results[1].responses[0].fullTextAnnotation;
		console.log(fullTextAnnotation.text);
	})
	
	.catch((err) => {
		console.error('ERRORes:', err);
	});
	
}

// [END functions_ocr_detect]

// [START functions_ocr_process]
/**
 * Cloud Function triggered by Cloud Storage when a file is uploaded.
 *
 * @param {object} event The Cloud Functions event.
 * @param {object} event.data A Google Cloud Storage File object.
 */
exports.processImage = function processImage (event) {
  
	let file = event.data;
	
	return Promise.resolve()
	
	.then(() => {
      if (file.resourceState === 'not_exists') {
        // This was a deletion event, we don't want to process this
        return;
      }

      if (!file.bucket) {
        throw new Error('Bucket not provided. Make sure you have a "bucket" property in your request');
      }
      if (!file.name) {
        throw new Error('Filename not provided. Make sure you have a "name" property in your request');
      }
	  
      //return detectText(storage.bucket(file.bucket).file(file.name));
	  return detectText(storage.bucket(file.bucket).file(file.name));
    })
    .then(() => {
      console.log(`File ${file.name} processed.`);
    });
	
};
// [END functions_ocr_process]
```

Both the Cloud Vision API and Cloud Functions API are enabled:

![apis](https://cloud.githubusercontent.com/assets/28631206/26278746/492202e2-3da2-11e7-96ef-bba47dd9d489.png)

**Comment by jmdobry** [comment]
What images are you uploading? Perhaps they are malformed?

**Comment by rafaelvizcaino** [comment]
@jmdobry , bullseye! The *.jpg extension is not (yet) accepted by Cloud Vision. I generated another image *.jpeg, now everything works perfectly. Many thanks to you and @jasonpolites for your support!

**Comment by shouse** [comment]
As a short follow up, I'm able to use a .jpg successfully.  I provide it as a local file and not on cloud storage, but should not make a difference.

**Comment by brahama** [comment]
Does that code still works? i am testing it in cloud functions and get errors for the sintax like:
const storage = require('@google-cloud/storage')();

TypeError: storage.bucket is not a function

The same for vision.readdocuemnt...

Thanks
